src/main/java/com/programmers/springbootbasic/presentation/validation/InputValidator.java

```diff
@@ -0,0 +1,38 @@
+package com.programmers.springbootbasic.presentation.validation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class InputValidator<T> {
+
+    protected final List<validator<T>> validators = new ArrayList<>();
+
+    @FunctionalInterface
+    public interface validator<T> {
+        String getErrorMessage(T val);
+    }
+
+    public Optional<List<String>> validate(T input) {
+        var errs = getError(input);
+        if(errs.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(errs);
+    }
+
+    public void addValidator(validator<T> validator) {
+        this.validators.add(validator);
+    }
+
+    private List<String> getError(T input) {
+        List<String> errorMessages = new ArrayList<>();
+        for (validator<T> validator : validators) {
+            String errorMessage = validator.getErrorMessage(input);
+            if (errorMessage != null) {
+                errorMessages.add(errorMessage);
+            }
+        }
+        return errorMessages;
+    }
```

Comment:
stream으로 작성하면 좀 더 가독성있게 작성할 수 있을 것 같습니다.

```suggestion
        return validators.stream()
                         .map(validator -> validator.getErrorMessage(input))
                         .filter(Optional::isPresent)
                         .map(Optional::get)
                         .collect(Collectors.toList());
```

Reply:
[a3e1cc9](https://github.com/prgrms-be-devcourse/springboot-basic/pull/902/commits/a3e1cc9a3cb63a03e3fff950475d347fa9c888bf)

오 ㅠㅠ 생각도 못했습니다!!! 지적 감사합니다!!!!
근데 저 부분에서 반환 값이 string 이거나 null 인데 옵셔널로 바꾸는게 더 낫다고 보시나요?

Reply:
제가 주환님 말씀을 잘 이해못했는데 List에 null 값이 포함될 수 있다는 말씀일까요?
isPresent를 통해 값이 존재하는지 확인하고 get으로 가져온다면 isPresent에서 필터링 되서 값은 String 값만 반환 받을 것 같습니다 🤔

Reply:
@zxcv9203  아 map 함수를 통해 반환 되는 값이 null 아니면 String 인 상황입니다! `getErrorMessage(input)` 함수 return 값이 Optional 이 아닌 상황인데, 혹시 Optional 로 바꾸는 것이 나을지... 에 대한 고민이었습니다! 말을 정돈되지 않게 말씀드린 점 죄송합니다!

Reply:
아 아닙니다 제가 잘못이해한거라 ㅋㅋㅋ 주환님 잘못 없습니다,,

제가 착각해서 Optional인줄 알았습니다 ㅎㅎ.. 오버헤드도 존재하고 Optional로 감쌀 필요는 없을 것 같습니다.

```java
        return validators.stream()
                .map(validator -> validator.getErrorMessage(input))
                .filter(Objects::nonNull)
                .toList();

```

Optional 사용관련하여 어떤 상황에서 올바른 사용일지에 대한 글도 도움이 될 것 같습니다..!

https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/
https://mangkyu.tistory.com/203